book: add tuning guide (#2548)
firedancer-io/firedancer: book/guide/tuning.md

# Performance Tuning

## Overview
The Firedancer validator is composed of a handful of threads, each
performing one of eleven distinct jobs. Some jobs only need one thread
to do them, but certain jobs require many threads performing the same
work in parallel.

Each thread is given a CPU core to run on, and threads take ownership of
the core: never sleeping or letting the operating system use it for
another purpose. The combination of a job, and the thread it runs on,
and the CPU core it is assigned to is called a tile. The eleven kinds of
tile are,

| Tile   | Description |
|--------|-------------|

> | `net`  | Responsible for sending and receiving network packets from the network device |

Reviewer: ```suggestion
| `net`  | Sends and receives network packets from the network device |
```

> | `quic` | A QUIC server for receiving transactions from clients. Performs all connection management and packet processing to manage and implement the QUIC protocol |

Reviewer: This is the only one that doesn't start with a verb.
Maybe "Receives transactions from clients, performing all ..."

| `verify` | Verifies the cryptographic signature of incoming transactions, filtering invalid ones |
| `dedup` | Checks for and filters out duplicated incoming transactions |
| `pack` | Collects incoming transactions and smartly schedules them for execution when we are leader |
| `bank` | Executes transactions that have been scheduled when we are leader |
| `poh`  | Continuously hashes in the background, and mixes the hash in with executed transactions to prove passage of time |
| `shred` | Distributes block data to the network when leader, and receives and retransmits block data when not leader |
| `store` | Receives block data when we are leader, or from other nodes when they are leader, and stores it locally in a database on disk |
| `metric` | Collects monitoring information about other tiles and serves it on a HTTP endpoint |
| `sign` | Holds the validator private key, and receives and responds to signing requests from other tiles |

> These tiles communicate with each other via. shared memory queues. The

Reviewer: As far as I know, via isn't an abbreviation, so it doesn't need a period

| `shred`  | 2               | Throughput is mainly dependent on cluster size, 2 tiles is enough to handle current `mainnet-beta` conditions. In benchmarking, if the cluster size is small, 1 tile can handle >1M TPS |

## Testing
Firedancer includes a simple benchmarking tool for measuring the
transaction throughput of the validator when it is leader, in
transactions per second (TPS). In practice, the Solana network
performance is limited by two factors that are unrelated to what
this tool measures:

 - The replay performance of the slowest nodes in the network, and if
they can keep up
 - The consensus limits on block size and data size

In particular, consensus limits on the Solana protocol limit the network
strictly to around 81,000 TPS. But the tool can be useful for testing
local affinity and layout configurations.

The benchmark runs on a single machine and performs the following:

 1. A new genesis is created, and set of accounts are pre-funded
 2. A set of CPU cores is assigned to generating and signing simple

> transfers between these accounts as fast as possible

Reviewer: ```suggestion
transactions using these accounts as fast as possible
```

 3. Another set of CPU cores is assigned to sending these transfers
via. QUIC over loopback to the locally running validator
 4. Around once a second, an RPC call is made to get the total count of

> transfers that have executed on the chain, and this information is

Reviewer: ```suggestion
transactions that have executed on the chain, and this information is
```

printed to the console

The benchmark is currently quite synthetic, as it only measures single
node performance, in an idealized case where all transactions are

> non-conflicting simple transfers.

Reviewer: ```suggestion
non-conflicting.
```
